### PR TITLE
Fix: Integrations should be closeable

### PR DESCRIPTION
**Pull request by rxlabz**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

- add a integration disposer list to `Hub`
- dispose integrations on `Hub.close` 
- update the isolateErrorIntegration to allow closing it's port 

## :bulb: Motivation and Context

We noticed that the sample Dart process was never terminated due to the isolateErrorIntegration receptor.

## :green_heart: How did you test it?
unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/isolate_error_integration.dart [hunk 1]
```diff
@@ -5,15 +5,17 @@ import 'protocol.dart';
 import 'sentry_options.dart';
 import 'throwable_mechanism.dart';
 
-/// integration that capture errors on the current Isolate Error handler
-/// which is the main thread.
-void isolateErrorIntegration(Hub hub, SentryOptions options) {
-  final receivePort = _createPort(hub, options);
+Integration initIsolateErrorIntegration(
```
Comment (marandaneto) [inline dart/lib/src/isolate_error_integration.dart]: on Android we implement `Closeable` https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/TempSensorBreadcrumbsIntegration.java#L23
which is the equivalent of this technique for `typedef` I'd say, sounds good.

Comment (marandaneto) [reply dart/lib/src/isolate_error_integration.dart]: we don't use `init` as a prefix for the other integrations, I'd rather stick to the normal naming

File: dart/lib/src/isolate_error_integration.dart [hunk 2]
```diff
@@ -5,15 +5,17 @@ import 'protocol.dart';
...
+  void Function(Function) getPortDisposer,
```
Comment (marandaneto) [inline dart/lib/src/isolate_error_integration.dart]: what's about a more generic name like `integrationDisposable` so we could reuse for the next integrations, could be a typedef and an optional parameter to the `integration` typedef

File: dart/lib/src/sentry.dart [hunk 3]
```diff
@@ -155,6 +162,8 @@ class Sentry {
     final hub = currentHub;
     _hub = NoOpHub();
     hub.close();
+
+    _disposeIsolate?.call();
```
Comment (marandaneto) [inline dart/lib/src/sentry.dart]: this should be in the `hub.close()` method, see:

https://github.com/getsentry/sentry-java/blob/2e3679d18ec8be209fb5ad1432fc26c4adf2a268/sentry/src/main/java/io/sentry/Hub.java#L278-L282

ideally, as List, other integrations might have the same problem

Comment (rxlabz) [reply dart/lib/src/sentry.dart]: ✅

File: dart/lib/src/isolate_error_integration.dart (hunk 1 above)
Comment (rxlabz) [reply dart/lib/src/isolate_error_integration.dart]: ✅

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,7 @@
 - Fix: NDK integration was being disabled by a typo
 - Fix: missing toList for debug meta #192
 - Enhancement: NavigationObserver to record Breadcrumbs for navigation events #197
+- Fix: close the isolateErrorIntegration receivePort on Sentry.close()
```
Comment (marandaneto) [inline CHANGELOG.md]: I'd change to something more generic now, which is a system to dispose integrations

File: dart/lib/src/hub_adapter.dart [hunk 5]
```diff
@@ -77,4 +77,7 @@ class HubAdapter implements Hub {
 
   @override
   SentryId get lastEventId => Sentry.lastEventId;
+
+  @override
+  void addIntegrationDisposer(disposer) {}
```
Comment (rxlabz) [inline dart/lib/src/hub_adapter.dart]: @marandaneto I'm not sure if this method must do something in the HubAdapter ?

File: dart/lib/src/sentry_options.dart [hunk 6]
```diff
@@ -338,7 +338,15 @@ typedef EventProcessor = FutureOr<SentryEvent> Function(
...
-typedef Integration = FutureOr<void> Function(Hub hub, SentryOptions options);
+typedef Integration = FutureOr<void> Function(
+  Hub hub,
+  SentryOptions options, [
+  AddIntegrationDisposer addDisposer,
+]);
+
+typedef IntegrationDisposer = void Function();
+
+typedef AddIntegrationDisposer = void Function(IntegrationDisposer);
```
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]: docs

File: dart/lib/src/hub.dart [hunk 7]
```diff
@@ -312,6 +321,11 @@ class Hub {
       }
     }
   }
+
+  /// add an integration disposer
+  /// disposers are called on [Hub.close]
+  void addIntegrationDisposer(IntegrationDisposer disposer) =>
+      _integrationDisposers.add(disposer);
```
Comment (marandaneto) [inline dart/lib/src/hub.dart]: why does it need to be a public method on the Hub, can we get rid of it somehow?

Comment (marandaneto) [reply dart/lib/src/hub.dart]: if we can't do it fully automatic, I'd rather prefer to add as named/positional parameter to the  `options.addIntegration` method

File: dart/lib/src/hub_adapter.dart (hunk 5 above)
Comment (marandaneto) [reply dart/lib/src/hub_adapter.dart]: I think this should not be part of the Hub at all.
See https://github.com/getsentry/sentry-dart/pull/209#discussion_r533322127

File: CHANGELOG.md (hunk 4 above)
Comment (rxlabz) [reply CHANGELOG.md]: ✅ `add an integration closing system`?

File: dart/lib/src/hub_adapter.dart (hunk 5 above)
Comment (rxlabz) [reply dart/lib/src/hub_adapter.dart]: ✅

**Review by bruno-garcia — APPROVED** [review]

**Pull request merged**